torchtune/models/llava_next/image_processor.py

```diff
@@ -0,0 +1,344 @@
+
+logger = logging.getLogger(__name__)
+
+ImageInput = Union["PIL.Image.Image", np.ndarray, "torch.Tensor"]
+
+
+class GetImagePatches(nn.Module):
+    def __init__(
+        self,
+        possible_resolutions: List[Tuple[int, int]],
+        patch_size: int,
+        resample: Union[str, torchvision.transforms.InterpolationMode] = "bilinear",
+    ) -> None:
+        super().__init__()
+        """
+        Class used to divide and image with variable resolutions into patches of equal sizes,
+        including the original image resized to [patch_size, patch_size].
+
+        Forward returns a tensor of shape [num_patches + 1, channels, patch_size, patch_size].
+
+        Args:
+            possible_resolutions (List[Tuple[int, int]]): List of possible resolutions as [height, width].
+            patch_size (int): Size of the patches to divide the image into.
+            resample (Union[str, torchvision.transforms.InterpolationMode]):
+            Resampling method, either 'bilinear', 'bicubic' or torchvision.transforms.InterpolationMode.
+        """
+
+        self.possible_resolutions = torch.tensor(possible_resolutions)
+
+        if isinstance(resample, torchvision.transforms.InterpolationMode):
+            self.resample = resample
+        if resample == "bilinear":
+            self.resample = torchvision.transforms.InterpolationMode.BILINEAR
+        elif resample == "bicubic":
+            self.resample = torchvision.transforms.InterpolationMode.BICUBIC
+        else:
+            raise ValueError(
+                "resample must be of type torchvision.transforms.InterpolationMode or ['bilinear', 'bicubic]."
+            )
+
+        self.center_crop = v2.CenterCrop(size=patch_size)
+
+        self.patch_size = patch_size
+
+    @staticmethod
+    def _get_new_size_without_distortion(
+        image_size: Tuple[int, int],
+        target_resolution: List[int],
+    ) -> Tuple[int, int]:
+
+        """
+        Determines the maximum dimensions to which an image can be resized without distorting its
+        aspect ratio, based on the target resolution.
+
+        Args:
+            image_size (Tuple[int, int]): The original dimensions of the image (height, width).
+            target_resolution (List[int]): The desired resolution to fit the image into (height, width).
+        Returns:
+            Tuple[int, int]: The optimal dimensions (height, width) to which the image should be resized.
+        Example:
+            >>> _get_new_size_without_distortion([200, 300], target_resolution = [450, 200])
+            (134, 200)
+            >>> _get_new_size_without_distortion([800, 600], target_resolution = [450, 1300])
+            (450, 338)
+        """
+
+        original_height, original_width = image_size
+        target_height, target_width = target_resolution
+
+        scale_w = target_width / original_width
+        scale_h = target_height / original_height
+
+        if scale_w < scale_h:
```

Comment:
nit: one quick comment explaining this would be helpful, explaining that you only resize to the shortest edge